### ISSUE TITLE
Replace `plan.Do` with `plan.Next`

### ISSUE THREAD

**Issue by coocood** [pingcap/tidb#107]
`plan.Do` method make it difficult to incrementally process table rows.

When `Do` method is executing, the only way to pause it is to use another goroutine with a channel, as currently the driver does.

And with Do method the only way to process data is to use an closure, but with next, we can process data without closure.

The proposed next method looks like:

```
`Next(ctx context.Context) (data []interface{}, rowkeys []*RowKeyEntry, err error)`
```

**Comment by siddontang** [comment]
+1
Use Next can remove closure which is hard to read, and at same time, we will avoid #57 finally.
to support this, we should maintain an iterator manually in group by, order by and distinct plan. 
maybe later we will add temporary table for these plans.

**Comment by shenli** [comment]
+1
Please add RowKeyList to the return value

**Comment by qiuyesuifeng** [comment]
+1
Do not like closure.

**Comment by coocood** [comment]
@shenli Updated

**Comment by cznic** [comment]
> When Do method is executing, the only way to pause it is to use another goroutine with a channel, as currently the driver does.

FTR: The `RowIterFunc` passed to `plan.Do` has two return values: `(more bool, err error)` and is called for every row the plan produces. When `RowIterFunc` returns `(false, nil)`, `plan.Do` returns immediately.

**Comment by siddontang** [comment]
Hi @cznic 

We know that return `(false, nil)` can terminate the Do immediately, but what we want to do is to traverse data iteratively and guarantee all queries in one transaction are handled in one goroutine.

Although we use channel + goroutine to support iteration in current implementation, a risk that the session is handled in two goroutines, we meet this panic, see #57.

Btw, maybe it is more convenient to write logic code with iteration (Next method, or even Previous method later) than callback with closure. :-) , e.g.

```
iter1 := NewIterator(result set1)
iter2 := NewIterator(result set2)

iter1.Next()
iter2.Next()

// get iter1 and iter2 data and do something, then iterate.  
iter1.Next()
iter2.Next()
```

**Comment by cznic** [comment]
The "extra" goroutine in `driver.go` is only a performance optimization. It is not necessary to do it that way. However, it enables to produce a result set in parallel/concurrently to consuming it.

**Comment by coocood** [comment]
@cznic You are right.
But if user want to scan through a very large table, fetching all rows before return to user may cause out of memory issue.

**Comment by cznic** [comment]
Of course. That's why the channel is buffered with a limited number of items. I think it's 1000 rows. IOW, *"fetching all rows before return to user"* does not apply. At most 1000 additional rows *may* get fetched in paralell to what the user/client already consumed. And the constant, 1000, can be tuned to a lower value probably without any serious performance degradartion.

Don't get me wrong, it's completely your decision how to design things. I'm just trying to keep you informed when I see a discussion based on assumptions which are not correct - to aid you, potentially, in making the above mentioned decision ;-)

**Comment by coocood** [comment]
@cznic 
I'm sorry for the mistaken statement I made above.

We actually implemented a server layer without the channel and goroutine which will be open-sourced soon.

The server layer indeed fetched all the rows before return to user by calling `Recordset.Rows` method.

And we want to access the transaction instance only in one goroutine to avoid concurrency issue.

And thank you for the information, it helps a lot.
